server/src-lib/Hasura/RQL/DDL/CustomTypes.hs

```diff
@@ -15,20 +15,26 @@ import qualified Database.PG.Query                 as Q
 import qualified Language.GraphQL.Draft.Syntax     as G
 
 import           Hasura.EncJSON
+import           Hasura.GraphQL.Validate.Types     (mkScalarTy)
 import           Hasura.Prelude
 import           Hasura.RQL.Types
 import           Hasura.SQL.Types
 
 import           Hasura.GraphQL.Schema.CustomTypes (buildCustomTypesSchemaPartial)
 
+-- | Validate the custom types and return any reused Postgres base types (as scalars)
 validateCustomTypeDefinitions
   :: (MonadValidate [CustomTypeValidationError] m)
-  => TableCache -> CustomTypes -> m ()
-validateCustomTypeDefinitions tableCache customTypes = do
+  => TableCache -- ^ The Postgres tables
+  -> CustomTypes -- ^ Custom types subject to validation
+  -> [PGScalarType] -- ^ List of all Postgres base types
+  -> m [PGScalarType] -- ^ List of reused Postgres base types (as scalars) in custom types
+validateCustomTypeDefinitions tableCache customTypes allPGScalars = do
   unless (null duplicateTypes) $ dispute $ pure $ DuplicateTypeNames duplicateTypes
   traverse_ validateEnum enumDefinitions
-  traverse_ validateInputObject inputObjectDefinitions
-  traverse_ validateObject objectDefinitions
+  inputObjectReusedPGScalars <- concat <$> traverse validateInputObject inputObjectDefinitions
+  objectReusedPGScalars <- concat <$> traverse validateObject objectDefinitions
```

Comment:
All this threading around of the collected scalars seems like it would be much easier to do using [a writer monad](https://hackage.haskell.org/package/mtl-2.2.2/docs/Control-Monad-Writer-Strict.html). And if you use a `HashSet PGScalarType` instead of a `[PGScalarType]`, you won’t have to worry about duplicates! So you could just write

```haskell
validateCustomTypeDefinitions tableCache customTypes allPGScalars = execWriterT do
  unless (null duplicateTypes) $ dispute $ pure $ DuplicateTypeNames duplicateTypes
  traverse_ validateEnum enumDefinitions
  traverse_ validateInputObject inputObjectDefinitions
  traverse_ validateObject objectDefinitions
```

and every time you find a reference to a scalar type, you could write `tell $ S.singleton reusedScalar`.